File: WatchPartyApp/watchparty-ui/src/services/youTubeService.ts
import { apiService } from './api';
import { YouTubeSearchResponse, YouTubeVideo } from '../types';

class YouTubeService {
    async searchVideos(query: string, maxResults: number = 10): Promise<YouTubeSearchResponse> {
        return apiService.get<YouTubeSearchResponse>(`/api/youtube/search?query=${encodeURIComponent(query)}&maxResults=${maxResults}`);
    }

    async getVideoDetails(videoId: string): Promise<YouTubeVideo> {
        return apiService.get<YouTubeVideo>(`/api/youtube/video/${videoId}`);
    }

    extractVideoIdFromUrl(url: string): string | null {
        if (!url) return null;

        // YouTube URL patterns
        const patterns = [
            /(?:youtube\.com\/watch\?v=|youtu\.be\/|youtube\.com\/embed\/)([a-zA-Z0-9_-]{11})/,
            /youtube\.com\/watch\?.*v=([a-zA-Z0-9_-]{11})/
        ];

        for (const pattern of patterns) {
            const match = url.match(pattern);
            if (match) {
                return match[1];
            }
        }

        return null;
    }

    isYouTubeUrl(url: string): boolean {
        if (!url) return false;
        return /(?:youtube\.com|youtu\.be)/i.test(url);
    }

    // Convert YouTube URL to embeddable format for video player
    convertToEmbedUrl(url: string): string {
        const videoId = this.extractVideoIdFromUrl(url);
        if (!videoId) return url;
        
        // For our video player, we'll use a direct YouTube embed URL
        // This will work with HTML5 video elements that support it
        return `https://www.youtube.com/embed/${videoId}?enablejsapi=1&origin=${window.location.origin}`;
    }

    // Get direct video stream URL (for HTML5 video elements)
    // Note: This is complex due to YouTube's restrictions. 
    // For now, we'll return the embed URL and handle it in the video player
    getVideoStreamUrl(url: string): string {
        return this.convertToEmbedUrl(url);
    }

    // Get YouTube thumbnail URL
    getThumbnailUrl(videoId: string, quality: 'default' | 'medium' | 'high' = 'medium'): string {
        const qualityMap = {
            default: 'default',
            medium: 'mqdefault',
            high: 'hqdefault'
        };
        
        return `https://img.youtube.com/vi/${videoId}/${qualityMap[quality]}.jpg`;
    }

    // Format duration from ISO 8601 format to readable format
    formatDuration(duration: string): string {
        try {
            // Handle ISO 8601 duration format (PT4M13S)
            const match = duration.match(/PT(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?/);
            if (!match) return duration;

            const hours = parseInt(match[1]) || 0;
            const minutes = parseInt(match[2]) || 0;
            const seconds = parseInt(match[3]) || 0;

            const parts = [];
            if (hours > 0) parts.push(`${hours}:${minutes.toString().padStart(2, '0')}`);
            else parts.push(minutes.toString());
            parts.push(seconds.toString().padStart(2, '0'));

            return parts.join(':');
        } catch {
            return duration;
        }
    }
}

export const youTubeService = new YouTubeService();

File: WatchPartyApp/WatchPartyApp/BusinessLogic/Services/YouTubeService.cs
using Google.Apis.Services;
using Google.Apis.YouTube.v3;
using System.Text.RegularExpressions;
using WatchPartyApp.BusinessLogic.Interfaces;
using WatchPartyApp.DTOs;
using System.Xml;
using System.Text.Json;

namespace WatchPartyApp.BusinessLogic.Services
{
    public class YouTubeService : IYouTubeService
    {
        private readonly Google.Apis.YouTube.v3.YouTubeService _googleYouTubeService;
        private readonly IConfiguration _configuration;
        private readonly ILogger<YouTubeService> _logger;
        private readonly HttpClient _httpClient;

        public YouTubeService(IConfiguration configuration, ILogger<YouTubeService> logger, HttpClient httpClient)
        {
            _configuration = configuration;
            _logger = logger;
            _httpClient = httpClient;

            var apiKey = configuration["YouTube:ApiKey"];
            if (string.IsNullOrEmpty(apiKey))
            {
                throw new InvalidOperationException("YouTube API key is not configured");
            }

            _googleYouTubeService = new Google.Apis.YouTube.v3.YouTubeService(new BaseClientService.Initializer()
            {
                ApiKey = apiKey,
                ApplicationName = "WatchPartyApp"
            });
        }

        public async Task<YouTubeSearchResponse> SearchVideosAsync(string query, int maxResults = 10, string? pageToken = null)
        {
            try
            {
                var searchListRequest = _googleYouTubeService.Search.List("snippet");
                searchListRequest.Q = query;
                searchListRequest.MaxResults = maxResults;
                searchListRequest.Type = "video";
                searchListRequest.VideoEmbeddable = SearchResource.ListRequest.VideoEmbeddableEnum.True__;
                searchListRequest.Order = SearchResource.ListRequest.OrderEnum.Relevance;

                if (!string.IsNullOrEmpty(pageToken))
                {
                    searchListRequest.PageToken = pageToken;
                }

                var searchListResponse = await searchListRequest.ExecuteAsync();

                var videos = new List<YouTubeVideoDto>();
                foreach (var searchResult in searchListResponse.Items)
                {
                    if (searchResult.Id.Kind == "youtube#video")
                    {
                        var video = new YouTubeVideoDto
                        {
                            VideoId = searchResult.Id.VideoId,
                            Title = searchResult.Snippet.Title,
                            Description = searchResult.Snippet.Description,
                            ThumbnailUrl = searchResult.Snippet.Thumbnails.Medium?.Url ?? searchResult.Snippet.Thumbnails.Default__.Url,
                            ChannelTitle = searchResult.Snippet.ChannelTitle,
                            PublishedAt = searchResult.Snippet.PublishedAtDateTimeOffset?.DateTime ?? DateTime.UtcNow
                        };

                        videos.Add(video);
                    }
                }

                // Get video durations
                if (videos.Any())
                {
                    await PopulateVideoDurations(videos);
                }

                return new YouTubeSearchResponse
                {
                    Videos = videos,
                    NextPageToken = searchListResponse.NextPageToken,
                    TotalResults = (int)(searchListResponse.PageInfo.TotalResults ?? 0)
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error searching YouTube videos");
                throw;
            }
        }

        public async Task<YouTubeVideoDto?> GetVideoDetailsAsync(string videoId)
        {
            try
            {
                var videosListRequest = _googleYouTubeService.Videos.List("snippet,contentDetails,statistics");
                videosListRequest.Id = videoId;

                var videosListResponse = await videosListRequest.ExecuteAsync();
                var video = videosListResponse.Items.FirstOrDefault();

                if (video == null)
                {
                    return null;
                }

                return new YouTubeVideoDto
                {
                    VideoId = video.Id,
                    Title = video.Snippet.Title,
                    Description = video.Snippet.Description,
                    ThumbnailUrl = video.Snippet.Thumbnails.Medium?.Url ?? video.Snippet.Thumbnails.Default__.Url,
                    ChannelTitle = video.Snippet.ChannelTitle,
                    Duration = ParseYouTubeDuration(video.ContentDetails.Duration),
                    PublishedAt = video.Snippet.PublishedAtDateTimeOffset?.DateTime ?? DateTime.UtcNow
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting YouTube video details");
                throw;
            }
        }

        public string? ExtractVideoIdFromUrl(string url)
        {
            if (string.IsNullOrWhiteSpace(url))
                return null;

            // YouTube URL patterns
            var patterns = new[]
            {
                @"(?:youtube\.com\/watch\?v=|youtu\.be\/|youtube\.com\/embed\/)([a-zA-Z0-9_-]{11})",
                @"youtube\.com\/watch\?.*v=([a-zA-Z0-9_-]{11})"
            };

            foreach (var pattern in patterns)
            {
                var match = Regex.Match(url, pattern, RegexOptions.IgnoreCase);
                if (match.Success)
                {
                    return match.Groups[1].Value;
                }
            }

            return null;
        }

        public bool IsYouTubeUrl(string url)
        {
            if (string.IsNullOrWhiteSpace(url))
                return false;

            return Regex.IsMatch(url, @"(youtube\.com|youtu\.be)", RegexOptions.IgnoreCase);
        }

        // Get direct video stream URL for YouTube videos
        public Task<string?> GetDirectVideoUrlAsync(string youtubeUrl)
        {
            try
            {
                var videoId = ExtractVideoIdFromUrl(youtubeUrl);
                if (string.IsNullOrEmpty(videoId))
                    return Task.FromResult<string?>(null);

                // Use youtube-dl-like approach to get direct video URLs
                // For now, we'll return a proxy URL that our backend can handle
                return Task.FromResult<string?>($"/api/youtube/stream/{videoId}");
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting direct video URL for YouTube video");
                return Task.FromResult<string?>(null);
            }
        }

        private async Task PopulateVideoDurations(List<YouTubeVideoDto> videos)
        {
            try
            {
                var videoIds = videos.Select(v => v.VideoId).ToList();
                var videosListRequest = _googleYouTubeService.Videos.List("contentDetails");
                videosListRequest.Id = string.Join(",", videoIds);

                var videosListResponse = await videosListRequest.ExecuteAsync();

                foreach (var video in videosListResponse.Items)
                {
                    var matchingDto = videos.FirstOrDefault(v => v.VideoId == video.Id);
                    if (matchingDto != null)
                    {
                        matchingDto.Duration = ParseYouTubeDuration(video.ContentDetails.Duration);
                    }
                }
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Error getting video durations");
                // Don't throw - durations are optional
            }
        }

        private static TimeSpan ParseYouTubeDuration(string duration)
        {
            try
            {
                // YouTube duration format: PT4M13S (4 minutes 13 seconds)
                return XmlConvert.ToTimeSpan(duration);
            }
            catch
            {
                return TimeSpan.Zero;
            }
        }

        protected virtual void Dispose(bool disposing)
        {
            if (disposing)
            {
                _googleYouTubeService?.Dispose();
            }
        }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
    }
}
